File: sdk-ts/src/dashboard/persistence.ts
/**
 * Dashboard notification for live call updates.
 *
 * When the SDK completes a call, it fires a POST to the standalone dashboard
 * (if running) so calls appear in real time.  Data lives only in memory —
 * nothing is written to disk.
 */

export function notifyDashboard(
  callData: Record<string, unknown>,
  port = 8000,
): void {
  try {
    const body = JSON.stringify(callData);
    const req = (require('node:http') as typeof import('node:http')).request(
      {
        hostname: '127.0.0.1',
        port,
        path: '/api/dashboard/ingest',
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          'Content-Length': Buffer.byteLength(body),
        },
        timeout: 1000,
      },
      () => { /* ignore response */ },
    );
    req.on('error', () => { /* dashboard not running, ignore */ });
    req.write(body);
    req.end();
  } catch {
    // silently ignore
  }
}

File: sdk-py/patter/dashboard/persistence.py
"""Dashboard notification for live call updates.

When the SDK completes a call, it fires a POST to the standalone dashboard
(if running) so calls appear in real time.  Data lives only in memory —
nothing is written to disk.
"""

from __future__ import annotations

__all__ = ["notify_dashboard"]

import dataclasses
import json
from typing import Any


def _default_serializer(obj: Any) -> Any:
    """JSON serializer that handles frozen dataclasses."""
    if dataclasses.is_dataclass(obj) and not isinstance(obj, type):
        return dataclasses.asdict(obj)
    raise TypeError(f"Object of type {type(obj)} is not JSON serializable")


def notify_dashboard(call_data: dict[str, Any], port: int = 8000) -> None:
    """Fire-and-forget POST to a running standalone dashboard.

    Silently ignores connection errors — the dashboard may not be running.
    """
    try:
        import httpx

        httpx.post(
            f"http://127.0.0.1:{port}/api/dashboard/ingest",
            json=json.loads(json.dumps(call_data, default=_default_serializer)),
            timeout=1.0,
        )
    except Exception:
        pass
